### ISSUE TITLE
Form Login does not redirect correctly to the login path

### ISSUE THREAD

**Issue by yellow1912** [symfony/symfony#23116]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | yes
| RFC?             | no
| Symfony version  | 3.3.0

I run into this weird issue today after upgrading from 3.1 to 3.3. My admin backend is /backend, when the user is not logged in he/she should be redirected to /backend/login. After the upgrade, the user is no longer redirected but 403 Access Denied Exception is thrown instead.

My current security settings:

```
firewalls:
	backend:
	    pattern: ^/([\w\d\-_]+/)?backend/
	    provider: acme_account_user_provider
	    form_login:
	        login_path:  acme.account.backend.account.login # this is /backend/login
	        check_path:  acme.account.backend.account.check # this is /backend/login_check
	#                default_target_path: /backend/
	        failure_path: /backend/login/
	        success_handler: acme.security.authentication.handler.success
	        failure_handler: acme.security.authentication.handler.failure
	        use_referer: true
	        username_parameter: 'acme_security[_username]'
	        password_parameter: 'acme_security[_password]'
	        csrf_parameter: 'acme_security[_token]' # define csrf form name
	        csrf_token_generator: security.csrf.token_manager # to check csrf if define csrf_provider
	        # always_use_default_target_path: true
	#                default_target_path: /backend/dashboard
	#                failure_path: /backend/login
	    remember_me:
	        secret: "%secret%"
	        remember_me_parameter: 'acme_security[_remember_me]'
	    logout:
	        path:   /backend/logout
	        target: /backend/login
	    security: true
	    anonymous: true
	    entry_point: acme.account.security.entry_point.login

access_control:
    - { path: '^/([\w\d\-_]+/)?backend/login',        role: IS_AUTHENTICATED_ANONYMOUSLY }
    - { path: '^/([\w\d\-_]+/)?backend/login_check',  role: IS_AUTHENTICATED_ANONYMOUSLY }
    - { path: '^/([\w\d\-_]+/)?backend/resetting',    role: IS_AUTHENTICATED_ANONYMOUSLY }
    - { path: '^/([\w\d\-_]+/)?backend',              role: ROLE_ACME_DEFAULT_ADMIN }
```

Stack trace:

> Symfony\Component\Security\Core\Exception\AccessDeniedException:
> Access Denied.
> 
>   at vendor/symfony/symfony/src/Symfony/Component/Security/Http/Firewall/AccessListener.php:70
>   at Symfony\Component\Security\Http\Firewall\AccessListener->handle(object(GetResponseEvent))
>      (vendor/symfony/symfony/src/Symfony/Component/Security/Http/Firewall.php:69)
>   at Symfony\Component\Security\Http\Firewall->onKernelRequest(object(GetResponseEvent))
>      (vendor/symfony/symfony/src/Symfony/Bundle/SecurityBundle/EventListener/FirewallListener.php:49)
>   at Symfony\Bundle\SecurityBundle\EventListener\FirewallListener->onKernelRequest(object(GetResponseEvent))
>      (local/cache/dev/nileadDevDebugProjectContainer.php:3321)
>   at nileadDevDebugProjectContainer->{closure}(object(GetResponseEvent), 'kernel.request', object(TraceableEventDispatcher))
>   at call_user_func(object(Closure), object(GetResponseEvent), 'kernel.request', object(TraceableEventDispatcher))
>      (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/WrappedListener.php:112)
>   at Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke(object(GetResponseEvent), 'kernel.request', object(ContainerAwareEventDispatcher))
>   at call_user_func(object(WrappedListener), object(GetResponseEvent), 'kernel.request', object(ContainerAwareEventDispatcher))
>      (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:174)
>   at Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(array(object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener), object(WrappedListener)), 'kernel.request', object(GetResponseEvent))
>      (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:43)
>   at Symfony\Component\EventDispatcher\EventDispatcher->dispatch('kernel.request', object(GetResponseEvent))
>      (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php:146)
>   at Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch('kernel.request', object(GetResponseEvent))
>      (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:129)
>   at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
>      (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:68)
>   at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
>      (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:171)
>   at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
>      (web/app_dev.php:27)

**Comment by yellow1912** [comment]
It's worth noting that the failure handler service is not triggered at all.

**Comment by xabbuh** [comment]
Can you give a small example project that makes it possible to reproduce your issue?

**Comment by nicolas-grekas** [comment]
Could be related to #23061 and #23027 ?

**Comment by yellow1912** [comment]
@nicolas-grekas  I tried to apply the changes in the above PRs but that did not fix the issue.

@xabbuh I will do some more debugging, if I cannot figure it out I will do as you suggest and regenerate a mini app to see if I have the same issue.

**Comment by yellow1912** [comment]
Not sure if this is intentional but:

In UsernamePasswordFormAuthenticationListener

```
protected function requiresAuthentication(Request $request)
    {
        if ($this->options['post_only'] && !$request->isMethod('POST')) {
            return false;
        }

        return parent::requiresAuthentication($request);
    }
```

By default post_only is true, so it means if I visit the backend/ page on GET request then the FormAuth will not be triggered.

The second thing I had to fix was to change check_path from backend/login_check to backend/

Not sure why my old settings worked for older version. I couldn't find anything in the changelogs that mention this. Perhaps my old settings were wrong to begin with, but I wonder why it worked.

**Comment by yellow1912** [comment]
Ok, So after further debugging, I think I'm seeing the issue here. With the above check in requiresAuthentication of FormLogin, when I go to backend I this AuthenticationListener will not be triggered. So it goes to AccessListener and there the exception is triggered (which seems to be correct).

Now what gets me confused is if in this case the FormLogin is passed, then how will the code knows it needs to redirect to 'login_path'. I would assume that some kind of exceptionListener must handle this part?

**Comment by yellow1912** [comment]
Ok, so I hunted down the issue (hopefully).

We have a custom ExceptionListener, in this ExceptionListener we would set our custom Response since we want to tweak the data we pass into the view a bit. After the upgrade, it seems like if you set the response in any of the listeners, the rest of the listeners will be skipped ( understandable).  So I added a check to skip handling AccessDeniedException in our custom listener like this:


```
if ($exception instanceof AccessDeniedException) {
            return;
        }
```

This is the only thing I had to do to fix our issue :).
